### PR TITLE
`plugin` MLFlow

### PR DESCRIPTION
**Pull request by zilto**
The MLFlow plugin for Hamilton includes two sets of features:
- Save and load machine learning models with the `MLFlowModelSaver` and `MLFlowModelLoader` materializers
- Automatically track data pipeline results in MLFlow with the `MLFlowTracker`.

## Changes
- `hamilton/plugins/mlflow_extensions.py` contains the materializers
- `hamilton/plugins/h_mlflow.py` contains the tracker
- `examples/mlflow/tutorial.ipynb` is a tutorial notebook

## How I tested this
- `tests/plugins/test_mlflow_extensions.py` tests the materializers
- currently no tests for the tracker

## TODO / ideas
- `display_all_functions()` and `visualize_execution()` from the `HamiltonTracker` to store in the MLFlow tracking server
- strong coupling between MLFlow experiments and runs with Hamilton UI projects and runs. The MLFlow UI allows for markdown fields which could contain a link to the Hamilton UI.
- log "datasets" used for tracking in the MLFlow UI. Accepts a "digest" which is equivalent to our fingerprinting concept
- log model input signatures
- add example on how to add hyperparameter tuning with nested runs
- explore the MLFlow features for LLMs and evaluation

### CODE REVIEW

File: examples/mlflow/README.md [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+[MLFlow](https://mlflow.org/) is an open-source Python framework for experiment tracking. It allows data science teams to store results, artifacts (machine learning models, figures, tables), and metadata in a principled way when executing data pipelines.
+
+The MLFlow plugin for Hamilton includes two sets of features:
+- Save and load machine learning models with the `MLFlowModelSaver` and `MLFlowModelLoader` materializers
+- Automatically track data pipeline results in MLFlow with the `MLFlowTracker`.
+
+This pairs nicely with the `HamiltonTracker` and the [Hamilton UI](https://hamilton.dagworks.io/en/latest/hamilton-ui/ui/) which gives you execution observability.
+
+We're looking forward to better link Hamilton "projects" with MLFlow "experiments" and runs from both projects.
```
Comment (skrawcz) [inline examples/mlflow/README.md:11]:
```suggestion
We're looking forward to the better linking of Hamilton "projects" with MLFlow "experiments" and runs from both projects.
```

File: examples/mlflow/README.md [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+    ```console
+    pip install -r requirements.txt
+    ```
+
+3. Explore the notebook `tutorial.ipynb`
+
+
+4. Launch the MLFlow user interface to explore results
+    ```console
+    mlflow ui
```
Comment (skrawcz) [inline examples/mlflow/README.md:29]: doesn't MLFlow need to be running first before the notebook?

File: examples/mlflow/README.md [hunk 3]
```diff
@@ -0,0 +1,35 @@
+# MLFLow plugin for Hamilton
+
+[MLFlow](https://mlflow.org/) is an open-source Python framework for experiment tracking. It allows data science teams to store results, artifacts (machine learning models, figures, tables), and metadata in a principled way when executing data pipelines.
+
+The MLFlow plugin for Hamilton includes two sets of features:
+- Save and load machine learning models with the `MLFlowModelSaver` and `MLFlowModelLoader` materializers
+- Automatically track data pipeline results in MLFlow with the `MLFlowTracker`.
+
+This pairs nicely with the `HamiltonTracker` and the [Hamilton UI](https://hamilton.dagworks.io/en/latest/hamilton-ui/ui/) which gives you execution observability.
```
Comment (skrawcz) [inline examples/mlflow/README.md:9]:
```suggestion
This pairs nicely with the `HamiltonTracker` and the [Hamilton UI](https://hamilton.dagworks.io/en/latest/hamilton-ui/ui/) which gives a better way to explore your code and understand assets/attributes produced, as well as execution observability.
```

File: hamilton/plugins/mlflow_extensions.py [hunk 4]
```diff
@@ -0,0 +1,187 @@
...
+                raise ValueError("Using `mode='registry` requires passing `model_name`")
+
+            if not (bool(self.version) ^ bool(self.version_alias)):
+                raise ValueError(
+                    "If using `mode='registry'` requires passing `version` OR `version_alias"
+                )
+
+            if self.version:
+                self.model_uri = f"models:/{self.model_name}/{self.version}"
+            elif self.version:
```
[the remark this answers is not in the document]
Comment (skrawcz) [reply hamilton/plugins/mlflow_extensions.py:142]: +1

File: hamilton/plugins/mlflow_extensions.py [hunk 5]
```diff
@@ -0,0 +1,187 @@
...
+                self.model_uri = f"models:/{self.model_name}@{self.version_alias}"
+
+    @classmethod
+    def name(cls) -> str:
+        return "mlflow"
+
+    @classmethod
+    def applicable_types(cls) -> Collection[Type]:
+        return [Any]
```
Comment (skrawcz) [inline hamilton/plugins/mlflow_extensions.py:151]: can this not be constrained?

**Review by skrawcz — COMMENTED** [review]
Only big one is `Any` typing. Would be nice to constrain it, but if it's too hard I guess that's the trade-off.

Otherwise minor things.

File: examples/mlflow/README.md (hunk 2 above)
Comment (zilto) [reply examples/mlflow/README.md:29]: Nope. It stores artifacts in metadata files that the UI server then reads (similar to my `ExperimentTracker`)

File: hamilton/plugins/h_mlflow.py [hunk 6]
```diff
@@ -0,0 +1,295 @@
...
+            # log materialized figure. Allows great degree of control over rendering format
+            # and also save interactive plotly visualization as HTML
+            elif node_tags["hamilton.data_saver.sink"] in ["plt", "plotly"]:
+                materializer_node = self.graph[node_name]
+                materialized_node = self.graph[materializer_node.required_dependencies.pop()]
+                figure = node_kwargs[materialized_node.name]
+
+                path = get_path_from_metadata(result)
+                if path:
+                    self.client.log_figure(self.run_id, figure, path)
```
Comment (elijahbenizzy) [inline hamilton/plugins/h_mlflow.py:226]: Hmm, this feels funky. Why are we logging both the figure and the path?

File: hamilton/plugins/mlflow_extensions.py [hunk 7]
```diff
@@ -0,0 +1,187 @@
...
+    """
+
+    model_uri: Optional[str] = None
+    mode: Literal["tracking", "registry"] = "tracking"
+    run_id: Optional[str] = None
+    path: Union[str, pathlib.Path] = "model"
+    model_name: Optional[str] = None
+    version: Optional[Union[str, int]] = None
+    version_alias: Optional[str] = None
+    flavor: Optional[str] = None
```
[the remark this answers is not in the document]
Comment (elijahbenizzy) [reply hamilton/plugins/mlflow_extensions.py:112]: If its just mlflow's name then that makes sense

File: examples/mlflow/tutorial.ipynb
[the remark this answers is not in the document]
Comment (zilto) [reply examples/mlflow/tutorial.ipynb:1]: Yeah, I added code snippets at the end for demonstration purposes. I turned them into markdown snippets

File: hamilton/plugins/h_mlflow.py (hunk 6 above)
Comment (zilto) [reply hamilton/plugins/h_mlflow.py:226]: This is the MLFlow API, not mine. You can read the line as 'write `figure` to `f"{self.run_id}/{path}"` '

File: hamilton/plugins/mlflow_extensions.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (zilto) [reply hamilton/plugins/mlflow_extensions.py:112]: This is the MLFlow terminology, didn't reinvent the wheel here

[the remark this answers is not in the document]
Comment (zilto) [reply hamilton/plugins/mlflow_extensions.py:112]: Each flavor is a library / backend which each have multiple types. We shouldn't explicitly handle them and let that be delegated to MLFlow

image

**Comment by zilto** [comment]
❗❗ I need to make a verification before merging. Thank you!

**Review by skrawcz — APPROVED** [review]
@zilto can you write a nice squash merge commit -- i.e. something to explain any choices/things you didn't add support for but could, etc.

**Pull request merged**